### PR TITLE
Add buffer_adaptor

### PR DESCRIPTION
**Pull request by Alex-PLACET**
Implement #46

### CODE REVIEW

File: include/sparrow/buffer_adaptor.hpp [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+#include "sparrow/buffer.hpp"
+#include "sparrow/contracts.hpp"
+
+namespace sparrow
+{
+    /**
+     * @brief Class which have internally a reference to a buffer<T> and provides the API of a buffer<U>
+     */
+    template <typename T, typename U>
+        requires(sizeof(U) < sizeof(T))
```
Comment (JohanMabille) [inline include/sparrow/buffer_adaptor.hpp:30]: Why this requirement?

File: include/sparrow/buffer_adaptor.hpp [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+#include <limits>
+#include <utility>
+
+#include "sparrow/buffer.hpp"
+#include "sparrow/contracts.hpp"
+
+namespace sparrow
+{
+    /**
+     * @brief Class which have internally a reference to a buffer<T> and provides the API of a buffer<U>
```
Comment (JohanMabille) [inline include/sparrow/buffer_adaptor.hpp:27]: From the implementation and the requirements below, I guess you meant it holds a a reference to a `buffer<U>` and provides an API to a `buffer<T>`? what do you think of renaming the template parameters into From and To to avoid any confusion?

File: include/sparrow/buffer_adaptor.hpp [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+        constexpr void push_back(const value_type& value);
+
+        constexpr void pop_back();
+
+        constexpr void resize(size_type new_size);
+        constexpr void resize(size_type new_size, const value_type& value);
+
+    private:
+
+        static constexpr double _sizeof_ratio = static_cast<double>(sizeof(buffer_reference_value_type))
```
Comment (JohanMabille) [inline include/sparrow/buffer_adaptor.hpp:126]: We should avoid names starting with `_`, these are reserved for the standard.

File: include/sparrow/buffer_adaptor.hpp [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+                                                / static_cast<double>(sizeof(value_type));
+
+        static constexpr auto _type_size = static_cast<size_type>(1. / _sizeof_ratio);
```
Comment (JohanMabille) [inline include/sparrow/buffer_adaptor.hpp:129]: What do you think of the names `from_to_size_ratio` and `to_from_size_ratio`?
Also why having a `size_type` and not a double here?

File: include/sparrow/buffer_adaptor.hpp [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+
+        buffer_reference& _buffer;
+    };
+
+    template <typename T, typename U>
+        requires(sizeof(U) < sizeof(T))
+    constexpr typename buffer_adaptor<T, U>::buffer_reference::size_type
+    buffer_adaptor<T, U>::index_for_buffer(size_type idx) const noexcept
+    {
+        return static_cast<buffer_reference::size_type>(static_cast<double>(idx) / _sizeof_ratio);
```
Comment (JohanMabille) [inline include/sparrow/buffer_adaptor.hpp:146]: - I think you can use `size_type` here.
- I think `idx * to_from_size_ratio` would be more efficient than a division.

File: include/sparrow/buffer_adaptor.hpp [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+        static constexpr auto _type_size = static_cast<size_type>(1. / _sizeof_ratio);
+
+        static constexpr size_type _extra_buffervalue_type_size = sizeof(buffer_reference_value_type)
+                                                                  % sizeof(value_type);
+
+        [[nodiscard]] constexpr buffer_reference::size_type index_for_buffer(size_type idx) const noexcept;
+
+        [[nodiscard]] buffer_reference::const_iterator get_buffer_reference_iterator(const_iterator pos);
+
+        buffer_reference& _buffer;
```
Comment (JohanMabille) [inline include/sparrow/buffer_adaptor.hpp:138]: naming convention: => `m_buffer`

File: include/sparrow/buffer_adaptor.hpp [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+
+    template <typename T, typename U>
+        requires(sizeof(U) < sizeof(T))
+    constexpr buffer_adaptor<T, U>::size_type buffer_adaptor<T, U>::size() const noexcept
+    {
+        const double new_size = static_cast<double>(_buffer.size()) * _sizeof_ratio;
+        SPARROW_ASSERT(
+            std::trunc(new_size) == new_size,
+            "The size of the buffer is not a multiple of the size of the new type"
+        );
```
Comment (JohanMabille) [inline include/sparrow/buffer_adaptor.hpp:303]: Should we express this as a requirement on the size of `T` and `U` instead? I don't see a case where we would convert a type into another with a non integer size of ratio.

File: include/sparrow/buffer_adaptor.hpp [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+    constexpr bool buffer_adaptor<T, U>::empty() const noexcept
+    {
+        return _buffer.empty();
+    }
+
+    template <typename T, typename U>
+        requires(sizeof(U) < sizeof(T))
+    constexpr void buffer_adaptor<T, U>::reserve(size_type new_cap)
+    {
+        _buffer.reserve(static_cast<size_type>(static_cast<double>(new_cap) / _sizeof_ratio));
```
Comment (JohanMabille) [inline include/sparrow/buffer_adaptor.hpp:332]: Same comment as for `index_for_buffer` implementation.

File: include/sparrow/buffer_adaptor.hpp (hunk 1 above)
Comment (Alex-PLACET) [reply include/sparrow/buffer_adaptor.hpp:30]: If sizeof(U) > sizeof(T), for example U is 8 bytes and T is 1 bytes.
If you want to insert/resize 2 T elements, you have to insert 1 U element.
So after inserting 2 T elements, the size of the buffer_adaptor is now 8. 
BTW, we should also have the requirement that sizeof(T) i a multiple of sizeof(U).

File: include/sparrow/buffer_adaptor.hpp (hunk 3 above)
Comment (Alex-PLACET) [reply include/sparrow/buffer_adaptor.hpp:126]: Ok

File: include/sparrow/buffer_adaptor.hpp (hunk 2 above)
Comment (Alex-PLACET) [reply include/sparrow/buffer_adaptor.hpp:27]: Good idea !

File: include/sparrow/buffer_adaptor.hpp (hunk 5 above)
Comment (Alex-PLACET) [reply include/sparrow/buffer_adaptor.hpp:146]: 👍

File: include/sparrow/buffer_adaptor.hpp (hunk 6 above)
Comment (Alex-PLACET) [reply include/sparrow/buffer_adaptor.hpp:138]: Done

File: include/sparrow/buffer_adaptor.hpp (hunk 4 above)
Comment (Alex-PLACET) [reply include/sparrow/buffer_adaptor.hpp:129]: Good idea

File: include/sparrow/buffer.hpp [hunk 9]
```diff
@@ -438,6 +443,15 @@ namespace sparrow
         get_data().p_end = copy_initialize(first, last, get_data().p_begin, get_allocator());
     }
 
+    template <class T>
+    template <class Range, allocator A>
```
Comment (Klaim) [inline include/sparrow/buffer.hpp:447]: Consider short-cutting to `template<std::ranges::input_range Range, allocator A>`

File: include/sparrow/buffer.hpp [hunk 10]
```diff
@@ -149,6 +150,10 @@ namespace sparrow
         template <class It, allocator A = allocator_type>
         constexpr buffer(It first, It last, const A& a = A());
 
+        template <class Range, allocator A = allocator_type>
```
Comment (Klaim) [inline include/sparrow/buffer.hpp:153]: Consider short-cutting to `template<std::ranges::input_range Range, allocator A>`

File: include/sparrow/buffer_adaptor.hpp [hunk 11]
```diff
@@ -0,0 +1,503 @@
...
+#include <limits>
+#include <utility>
+
+#include "sparrow/buffer.hpp"
+#include "sparrow/contracts.hpp"
+
+namespace sparrow
+{
+    /**
+     * @brief Class which have internally a reference to a buffer<From> and provides the API of a buffer<To>
```
Comment (Klaim) [inline include/sparrow/buffer_adaptor.hpp:27]: Remember that we rely on autobrief

File: include/sparrow/buffer_adaptor.hpp [hunk 12]
```diff
@@ -0,0 +1,503 @@
...
+            "The size of the buffer is not a multiple of the size of the new type"
+        );
+        return static_cast<size_type>(new_size);
+    }
+
+    template <typename To, typename From>
+        requires(sizeof(From) < sizeof(To))
+    constexpr typename buffer_adaptor<To, From>::size_type buffer_adaptor<To, From>::max_size() const noexcept
+    {
+        return static_cast<size_type>(static_cast<double>(m_buffer.max_size()) * m_from_to_size_ratio);
```
Comment (Klaim) [inline include/sparrow/buffer_adaptor.hpp:311]: Could you isolate/factorize that calculation so that it's not repeated in `capacity()` etc.? From what I understand only the initial value is a paremeter.

File: include/sparrow/buffer_adaptor.hpp [hunk 13]
```diff
@@ -0,0 +1,503 @@
...
+    }
+
+    template <typename To, typename From>
+        requires(sizeof(From) < sizeof(To))
+    constexpr typename buffer_adaptor<To, From>::iterator buffer_adaptor<To, From>::erase(const_iterator pos)
+    {
+        SPARROW_ASSERT_TRUE(cbegin() <= pos);
+        SPARROW_ASSERT_TRUE(pos <= cend());
+        const difference_type index = std::distance(cbegin(), pos);
+        const auto idx_for_buffer = index_for_buffer(static_cast<size_type>(index));
```
Comment (Klaim) [inline include/sparrow/buffer_adaptor.hpp:429]: When calling this function for immediate use in a buffer, consider adding assertions to check that index's bounds. (same in other functions)

File: test/test_buffer_adaptor.cpp [hunk 14]
```diff
@@ -0,0 +1,610 @@
...
+#include "sparrow/buffer_adaptor.hpp"
+
+#include "doctest/doctest.h"
+
+namespace sparrow
+{
+
+    TEST_SUITE("buffer_adaptor")
+    {
+        const std::array<uint8_t, 8> input{1u, 2u, 3u, 4u, 5u, 6u, 7u, 8u};
```
Comment (Klaim) [inline test/test_buffer_adaptor.cpp:27]: Also test with zero-sized array?

File: include/sparrow/buffer_adaptor.hpp (hunk 7 above)
Comment (JohanMabille) [reply include/sparrow/buffer_adaptor.hpp:303]: Indeed, my assumption was that something stored as uint8_t and that we cast in another type would actually be stored as uint8_t for technical reasons (dynamic type, deserialization), and that such a case would never happen. But that mmight not be the case, and anyway, we want to catch this kind of bugs.

File: include/sparrow/buffer_adaptor.hpp (hunk 1 above)
Comment (JohanMabille) [reply include/sparrow/buffer_adaptor.hpp]: Indeed, but I wonder if we should not add some constraints on the number of new elements we can insert / push_back etc in that case. That goes beyond our use case though (we will mostly convert from uint8_t to anything else), so let's keep it as is for now.

**Review by JohanMabille — APPROVED** [review]
LGTM once @Klaim  's comments have been addressed.

File: include/sparrow/buffer_adaptor.hpp (hunk 13 above)
Comment (Alex-PLACET) [reply include/sparrow/buffer_adaptor.hpp:439]: Done

File: include/sparrow/buffer_adaptor.hpp (hunk 12 above)
Comment (Alex-PLACET) [reply include/sparrow/buffer_adaptor.hpp]: I modified the code to avoid this long line of code

File: include/sparrow/buffer_adaptor.hpp (hunk 11 above)
Comment (Alex-PLACET) [reply include/sparrow/buffer_adaptor.hpp]: Done

File: include/sparrow/buffer.hpp (hunk 10 above)
Comment (Alex-PLACET) [reply include/sparrow/buffer.hpp]: Done

File: include/sparrow/buffer.hpp (hunk 9 above)
Comment (Alex-PLACET) [reply include/sparrow/buffer.hpp]: Done

File: include/sparrow/buffer_adaptor.hpp [hunk 15]
```diff
@@ -0,0 +1,526 @@
...
+                                                       / static_cast<double>(sizeof(value_type));
+
+        static constexpr size_type m_to_from_size_ratio = static_cast<size_type>(1. / m_from_to_size_ratio);
+
+        [[nodiscard]] constexpr buffer_reference::size_type index_for_buffer(size_type idx) const noexcept;
+
+        [[nodiscard]] buffer_reference::const_iterator get_buffer_reference_iterator(const_iterator pos);
+
+        buffer_reference& m_buffer;
+        const size_type m_max_size;
```
Comment (Klaim) [inline include/sparrow/buffer_adaptor.hpp:133]: The `const` makes this type non-copyable/movable. Do we really want that?
If yes, then add copy/move operations explicitly deleted so that it's clear to the user.

Comment (Klaim) [reply include/sparrow/buffer_adaptor.hpp:133]: Note also the tradefoff here of using memory instead of quick calculation (from a function), I'm not sure what's the best for our users. But it's easy to change later if needed (we are not ABI-sensible).

**Review by Klaim — APPROVED** [review]
LGTM once the `const` member thing is clarified 👍🏽

File: include/sparrow/buffer_adaptor.hpp (hunk 15 above)
Comment (Alex-PLACET) [reply include/sparrow/buffer_adaptor.hpp]: I made the member mutable 👍
